### ISSUE TITLE
Webpack 4 irregularly spikes CPU at 93% "after chunk asset optimization" during hot reload rebuild, CPU loads ~1.25 cores, sometimes Node dies with out-of-memory error

### ISSUE THREAD

**Issue by sompylasar** [webpack/webpack#6812]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

Webpack 4 irregularly spikes CPU at 93% "after chunk asset optimization" during hot reload rebuild, CPU loads ~1.25 cores, sometimes Node dies with out-of-memory error.

This happens after hot reload rebuild kicks in, during active development, so maybe there are some race conditions that kill it.

Usually this step completes in about 3 seconds or less:
```
[2018-03-21 12:45:58.473 AM PDT] [LOG]   [webpack/configDevServer] building... 🕚  93% after chunk asset optimization

[2018-03-21 12:46:11.969 AM PDT] [LOG]   [webpack/configDevServer] building... 🕚  93% asset optimization
```
```
[2018-03-21 12:46:17.145 AM PDT] [LOG]   [webpack/configDevClient] building... 🕚  93% after chunk asset optimization

[2018-03-21 12:46:30.240 AM PDT] [LOG]   [webpack/configDevClient] building... 🕚  93% asset optimization
```

Sometimes it spikes (10 seconds and more):
```
[2018-03-21 12:49:43.383 AM PDT] [LOG]   [webpack/configDevClient] building... 🕚  93% after chunk asset optimization

[2018-03-21 12:49:57.680 AM PDT] [LOG]   [webpack/configDevClient] building... 🕚  93% asset optimization
```

I managed to capture a couple `node --inspect` CPU profiler snapshots when it was in "after chunk asset optimization".

![cpu-20180321t000640_1](https://user-images.githubusercontent.com/498274/37698133-528d0e30-2c9e-11e8-891f-5a1b5dd5d56b.png)
[CPU-20180321T000640.cpuprofile.txt](https://github.com/webpack/webpack/files/1832334/CPU-20180321T000640.cpuprofile.txt)

![cpu-20180321t000542_1](https://user-images.githubusercontent.com/498274/37698143-5e5ca540-2c9e-11e8-9198-0a4e52ba2612.png)
[CPU-20180321T000542.cpuprofile.txt](https://github.com/webpack/webpack/files/1832335/CPU-20180321T000542.cpuprofile.txt)

**If the current behavior is a bug, please provide the steps to reproduce.**

The configuration is large and closed-source, sorry. It's a React-based project of ~1200 ES2017 JS files transpiled with Babel 6, ~340 SCSS files, all bundled with Webpack 4 parallelized with HappyPack.

I can help with testing new versions on this configuration if you happen to find a bottleneck in the CPU profiler snapshots I provided.

**What is the expected behavior?**

The "after chunk asset optimization" completes at about the same time as during the initial build.

~~**If this is a feature request, what is motivation or use case for changing the behavior?**~~

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

```
webpack 4.1.1
babel-core 6.26.0
node-sass 4.5.3
node 8.7.0
npm 5.6.0
macOS 10.12: Darwin __CENSORED__ 16.0.0 Darwin Kernel Version 16.0.0: Mon Aug 29 17:56:20 PDT 2016; root:xnu-3789.1.32~3/RELEASE_X86_64 x86_64
```

<details>
<summary><code>package.json</code> dependencies</summary>

```
  "dependencies": {
    "archiver": "^2.1.1",
    "async": "^2.5.0",
    "autoprefixer": "^7.1.4",
    "babel-core": "^6.21.0",
    "babel-plugin-dynamic-import-node": "^1.0.1",
    "babel-plugin-react-transform": "^2.0.2",
    "babel-plugin-syntax-dynamic-import": "^6.18.0",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-polyfill": "^6.20.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-react": "^6.16.0",
    "babel-preset-stage-1": "^6.16.0",
    "babel-register": "^6.18.0",
    "body-parser": "^1.14.1",
    "bootstrap-sass": "^3.3.5",
    "bugsnag": "^1.7.0",
    "chalk": "^2.3.1",
    "classnames": "^2.2.1",
    "codemirror": "^5.29.0",
    "color": "^2.0.1",
    "compression": "^1.6.0",
    "console-stamp": "^0.2.5",
    "contentful": "4.6.1",
    "cookie": "^0.3.1",
    "cookie-dough": "^0.1.0",
    "cookie-parser": "^1.4.0",
    "cross-env": "^5.1.3",
    "crypto": "^1.0.1",
    "csv-parse": "^1.2.0",
    "debug": "^3.1.0",
    "dotenv": "^4.0.0",
    "express": "^4.13.3",
    "font-awesome": "^4.7.0",
    "formidable": "^1.0.17",
    "froala-editor": "^2.6.5",
    "glob": "^7.1.1",
    "gm": "^1.23.0",
    "hellosign-sdk": "^1.1.14",
    "helmet": "^3.8.2",
    "history": "^3.3.0",
    "http-assert": "^1.2.0",
    "http-errors": "^1.5.0",
    "http-proxy": "^1.12.0",
    "isomorphic-style-loader": "^4.0.0",
    "jquery": "^3.2.1",
    "json2csv": "^3.11.2",
    "kickbox": "^2.0.0",
    "lodash": "^4.16.3",
    "mkdirp": "^0.5.1",
    "moment": "^2.13.0",
    "mongoose": "^4.13.0",
    "node-sass": "^4.5.0",
    "nodemailer": "^4.1.1",
    "nodemailer-mandrill-transport": "^1.2.0",
    "normalize.css": "^7.0.0",
    "npm-run-all": "^4.1.2",
    "nuka-carousel": "^3.0.0",
    "objectid": "^3.2.1",
    "pretty-error": "^2.1.0",
    "promise-throttle": "^0.3.1",
    "prop-types": "^15.5.8",
    "query-string": "^5.0.0",
    "raf": "^3.2.0",
    "react": "^16.0.0",
    "react-atellier": "^0.1.11",
    "react-contenteditable": "^2.0.2",
    "react-datepicker": "^1.0.4",
    "react-dom": "^16.0.0",
    "react-froala-wysiwyg": "^2.7.0",
    "react-height": "^3.0.0",
    "react-helmet": "^5.0.3",
    "react-input-range": "^1.1.2",
    "react-loadable": "^5.3.1",
    "react-measure": "^2.0.2",
    "react-motion": "^0.4.8",
    "react-places-autocomplete": "^5.0.0",
    "react-portal": "^3.0.0",
    "react-portal-tooltip": "^1.1.4",
    "react-pure-render": "^1.0.2",
    "react-redux": "^4.4.8",
    "react-router": "^3.0.5",
    "react-router-redux": "^2.1.0",
    "react-sortable-hoc": "^0.6.7",
    "react-stickynode": "^1.3.1",
    "redis": "^2.8.0",
    "redux": "^3.5.2",
    "redux-form": "^5.3.6",
    "redux-logger": "^2.3.2",
    "redux-saga": "^0.11.1",
    "rimraf": "^2.6.1",
    "s3": "^4.4.0",
    "sanitize-filename": "^1.6.1",
    "scss-to-json": "^1.1.0",
    "serialize-error": "^2.1.0",
    "serve-favicon": "^2.3.0",
    "stripe": "^4.9.0",
    "superagent": "^3.6.1",
    "svg-path-generator": "^1.1.0",
    "ua-parser-js": "^0.7.10",
    "url": "^0.11.0",
    "uuid": "^3.0.1",
    "verge": "^1.10.2",
    "when": "^3.7.7"
  },
  "devDependencies": {
    "ast-types": "^0.9.11",
    "babel-eslint": "^8.0.1",
    "babel-loader": "^7.1.4",
    "babel-plugin-transform-react-jsx-source": "^6.22.0",
    "bootstrap-sass-loader": "^1.0.10",
    "chai": "^4.1.0",
    "chai-xml": "^0.3.1",
    "css-loader": "^0.28.11",
    "escomplex": "^2.0.0-alpha",
    "eslint": "^4.8.0",
    "eslint-config-airbnb": "^15.1.0",
    "eslint-import-resolver-node": "^0.3.0",
    "eslint-import-resolver-webpack": "^0.8.1",
    "eslint-loader": "^2.0.0",
    "eslint-plugin-import": "^2.2.0",
    "eslint-plugin-jsx-a11y": "^5.1.1",
    "eslint-plugin-react": "^7.4.0",
    "file-loader": "^1.1.11",
    "font-awesome-sass-loader": "^2.0.1",
    "happypack": "^5.0.0-beta.1",
    "husky": "^0.14.3",
    "leasot": "^4.10.1",
    "listr": "^0.12.0",
    "minimist": "^1.2.0",
    "mocha": "^4.0.1",
    "nodemon": "^1.11.0",
    "postcss-loader": "^2.1.2",
    "prettyjson": "^1.2.1",
    "react-transform-catch-errors": "^1.0.2",
    "react-transform-hmr": "^1.0.4",
    "redbox-react": "^1.5.0",
    "redux-devtools": "^3.0.0-beta-3",
    "redux-devtools-dock-monitor": "^1.0.0-beta-3",
    "redux-devtools-log-monitor": "^1.0.0-beta-3",
    "resolve-url-loader": "^2.3.0",
    "sass-lint": "^1.10.2",
    "sass-loader": "^6.0.7",
    "sasslint-webpack-plugin": "^1.0.4",
    "sinon": "^4.0.0",
    "style-loader": "^0.20.3",
    "supertest": "^3.0.0",
    "svg-inline-loader": "^0.8.0",
    "svg-url-loader": "^2.3.2",
    "uglifyjs-webpack-plugin": "^1.2.4",
    "url-loader": "^1.0.1",
    "webpack": "^4.1.1",
    "webpack-bundle-analyzer": "^2.11.1",
    "webpack-dev-middleware": "^3.0.1",
    "webpack-hot-middleware": "^2.21.2",
    "webpack-node-externals": "^1.6.0"
  },
```

</details>

-------

Relevant:
- https://github.com/webpack/webpack/issues/1914 Process out of memory - Webpack
- https://github.com/webpack/webpack/issues/4558 large webpack build almost hangs at 91% on an "additional asset processing" step
- https://github.com/webpack/webpack/issues/537 Compile stuck at optimize chunk assets

**Comment by sompylasar** [comment]
Here's the webpack config extracted from runtime:
```js
{ mode: 'development',
     target: 'web',
     node: undefined,
     context: '/__CENSORED__',
     entry: 
      { client: 
         [ 'webpack-hot-middleware/client?path=/web-api/__webpack_hmr',
           'bootstrap-sass-loader!./src/webapp/rsh-web-theme/bootstrap.config.js',
           'font-awesome-sass-loader!./src/webapp/rsh-web-theme/font-awesome.config.js',
           './src/webapp/webapp.js' ] },
     output: 
      { path: '/__CENSORED__/static/dist',
        publicPath: '/dist/__webpack_hmr/',
        filename: '[name].js',
        chunkFilename: '[name].js',
        crossOriginLoading: 'anonymous',
        hashDigestLength: 40 },
     module: 
      { rules: 
         [ { test: /\.js$/,
             exclude: /node_modules/,
             use: [ { loader: 'happypack/loader' } ] },
           { test: /\.scss$/,
             use: 
              [ { loader: 'isomorphic-style-loader',
                  options: { sourceMap: true } },
                { loader: 'css-loader',
                  options: 
                   { modules: true,
                     importLoaders: 2,
                     sourceMap: true,
                     localIdentName: '[name]__[local]' } },
                { loader: 'postcss-loader',
                  options: { sourceMap: true, plugins: [Function: plugins] } },
                { loader: 'resolve-url-loader', options: { sourceMap: true } },
                { loader: 'sass-loader',
                  options: { outputStyle: 'expanded', sourceMap: true } } ] },
           { test: /\.css$/,
             use: 
              [ { loader: 'isomorphic-style-loader',
                  options: { sourceMap: true } },
                { loader: 'css-loader', options: { sourceMap: true } } ] },
           { test: /\.woff(\?v=\d+\.\d+\.\d+)?$/,
             loader: 'url-loader',
             options: 
              { emitFile: true,
                limit: 10000,
                mimetype: 'application/font-woff' } },
           { test: /\.woff2(\?v=\d+\.\d+\.\d+)?$/,
             loader: 'url-loader',
             options: 
              { emitFile: true,
                limit: 10000,
                mimetype: 'application/font-woff' } },
           { test: /\.ttf(\?v=\d+\.\d+\.\d+)?$/,
             loader: 'url-loader',
             options: 
              { emitFile: true,
                limit: 10000,
                mimetype: 'application/octet-stream' } },
           { test: /\.eot(\?v=\d+\.\d+\.\d+)?$/,
             loader: 'file-loader',
             options: { emitFile: true } },
           { test: /\/glyphicons-halflings-regular\.svg$/,
             loader: 'svg-url-loader',
             options: { emitFile: true, limit: 10000, noquotes: true } },
           { test: /\/fontawesome-webfont.svg(\?v=\d+\.\d+\.\d+)?$/,
             loader: 'svg-url-loader',
             options: { emitFile: true, limit: 10000, noquotes: true } },
           { test: /\/img\/(?![^\/]+?\.url\.svg)[^\/]+?\.svg$/,
             loader: 'svg-inline-loader?classPrefix=svg-[hash:base64:7]__&idPrefix=svg-[hash:base64:7]__' },
           { test: /\/img\/[^\/]+?\.url\.svg$/,
             loader: 'svg-url-loader',
             options: { emitFile: true, limit: 10000, noquotes: true } },
           { test: /\.(jpeg|jpg|png|gif)$/,
             loader: 'url-loader',
             options: { emitFile: true, limit: 2000 } } ] },
     externals: [],
     resolve: 
      { alias: 
         { 'normalize.css': '/__CENSORED__/node_modules/normalize.css/normalize.css' },
        modules: 
         [ '/__CENSORED__/src/api',
           '/__CENSORED__/src/webapp',
           '/__CENSORED__/src/common',
           'node_modules' ],
        extensions: [ '.js' ] },
     plugins: 
      [ ProgressPlugin {
          profile: undefined,
          handler: [Function],
          apply: [Function: buildProgressWebpackPluginApply] },
        DefinePlugin {
          definitions: 
           { __CLIENT__: true,
             __SERVER__: false,
             __DEVELOPMENT__: true,
             'process.env': { NODE_ENV: '"development"' } } },
        HappyPlugin {
          name: 'HappyPack',
          state: 
           { loaders: [],
             baseLoaderRequest: '',
             foregroundThreadPool: null,
             verbose: false,
             debug: false },
          config: 
           { id: '1',
             compilerId: 'default',
             threads: 3,
             verbose: true,
             debug: false,
             loaders: 
              [ { path: 'babel-loader',
                  query: 
                   { presets: 
                      [ [ 'env', { targets: { browsers: '> 0.5%' }, modules: false } ],
                        'react',
                        'stage-1' ],
                     plugins: 
                      [ 'syntax-dynamic-import',
                        'transform-decorators-legacy',
                        'transform-react-jsx-source',
                        '/__CENSORED__/webpack/babelPluginReactLoadable',
                        [ 'react-transform',
                          { transforms: 
                             [ { transform: 'react-transform-catch-errors',
                                 imports: [ 'react', 'redbox-react' ] },
                               { transform: 'react-transform-hmr',
                                 imports: [ 'react' ],
                                 locals: [ 'module' ] } ] } ] ],
                     ignore: [ 'build/**/*.js', 'node_modules' ],
                     babelrc: false,
                     comments: true,
                     forceEnv: 'webpack',
                     compact: false } },
                { path: 'eslint-loader' } ] },
          id: '1' },
        IgnorePlugin {
          resourceRegExp: /\bvertx\b/,
          contextRegExp: undefined,
          checkIgnore: [Function: bound checkIgnore] },
        BundleAnalyzerPlugin {
          opts: 
           { analyzerMode: 'server',
             analyzerHost: '127.0.0.1',
             analyzerPort: 8888,
             reportFilename: 'webpack-bundle-analyzer_production_browser-side.html',
             defaultSizes: 'parsed',
             openAnalyzer: false,
             generateStatsFile: false,
             statsFilename: 'stats.json',
             statsOptions: null,
             logLevel: 'info',
             startAnalyzer: true },
          server: null,
          logger: Logger { activeLevels: Set { 'info', 'warn', 'error', 'silent' } } },
        LoaderOptionsPlugin {
          options: 
           { minimize: false,
             debug: true,
             options: {},
             test: { test: [Function: test] } } },
        { apply: [Function: bound apply] },
        HotModuleReplacementPlugin {
          options: {},
          multiStep: undefined,
          fullBuildTimeout: 200,
          requestTimeout: 10000 } ],
     devtool: 'source-map',
     optimization: 
      { noEmitOnErrors: true,
        concatenateModules: false,
        namedModules: true,
        splitChunks: { chunks: 'all' },
        runtimeChunk: { name: 'webpack_runtime' },
        minimize: false,
        minimizer: [] } }
```

**Comment by sokra** [comment]
probably fixed by https://github.com/webpack/webpack-sources/pull/23

**Comment by sompylasar** [comment]
@sokra Thanks, makes sense. Link to the original report there for the record: https://github.com/webpack/webpack-sources/issues/29

When do you think I'll be able to test this in a webpack version?

Just got this which is quite frustrating:
```
[2018-03-21 01:46:14.945 AM PDT] [LOG]     [webpack/configDevServer] building... 🕚  92% chunk asset optimization

[2018-03-21 01:46:14.945 AM PDT] [LOG]     [webpack/configDevServer] building... 🕚  93% after chunk asset optimization

<--- Last few GCs --->

[41589:0x103001200]  2659129 ms: Mark-sweep 1417.1 (1729.2) -> 1417.1 (1729.2) MB, 1672.0 / 0.0 ms  allocation failure GC in old space requested
[41589:0x103001200]  2660363 ms: Mark-sweep 1417.1 (1729.2) -> 1417.1 (1674.7) MB, 1232.2 / 0.1 ms  last resort GC in old space requested
[41589:0x103001200]  2661641 ms: Mark-sweep 1417.1 (1674.7) -> 1417.1 (1662.2) MB, 1278.1 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0xd5f624a5e71 <JSObject>
    2: _parseMappings(aka SourceMapConsumer_parseMappings) [/__CENSORED__/node_modules/source-map/lib/source-map-consumer.js:508] [bytecode=0xd5f8bf3a879 offset=204](this=0xd5f1b30b671 <BasicSourceMapConsumer map = 0xd5fd3e072c9>,aStr=0xd5f1b30b651 <String[7]: AA6Q6CA>,aSourceRoot=0xd5fdbe82201 <null>)
    4: get [/__CENSORED__/node_module...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 4: v8::internal::Factory::NewFixedArray(int, v8::internal::PretenureFlag) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 5: v8::internal::HashTable<v8::internal::StringTable, v8::internal::StringTableShape>::EnsureCapacity(v8::internal::Handle<v8::internal::StringTable>, int, v8::internal::PretenureFlag) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 6: v8::internal::StringTable::LookupKey(v8::internal::Isolate*, v8::internal::StringTableKey*) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 7: v8::internal::StringTable::LookupString(v8::internal::Isolate*, v8::internal::Handle<v8::internal::String>) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 8: v8::internal::Runtime_KeyedGetProperty(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/ivanbabak/.nvm/versions/node/v8.7.0/bin/node]
 9: 0x199979a8463d
```

**Comment by sompylasar** [comment]
@sokra Also, why is the situation getting worse the more hot reloads happen? Is there a memory leak somewhere?

**Comment by zlk89** [comment]
Having same issue here. I have to disable parallel to avoid out of memory.

**Comment by xhite** [comment]
Same issue here

**Comment by tomasztomys** [comment]
Same issues here (at 5 different computers, 3 different projects)

**Comment by evilebottnawi** [comment]
Somebody can create minimum reproducible test repo?

**Comment by foisonocean** [comment]
Any update?

**Comment by xhite** [comment]
I wonder if it's rather related to webpack-dev-server. I switched to webpack-serve. No issues now

**Comment by rpellerin** [comment]
4.6.0 does not fix the issue for me. Still affected.

**Comment by evilebottnawi** [comment]
@rpellerin do you add `@import "webfont.scss";` in some scss file?

**Comment by abruzzihraig** [comment]
Still having the same issue at 4.6.0.

**Comment by sylvesteraswin** [comment]
I am having the same issue when import `ES6` compiled using `babel@7` and building using `webpack@4`

**Comment by sibelius** [comment]
can somebody review and merge this https://github.com/webpack/webpack-sources/pull/23

**Comment by jeffreybrowning** [comment]
@sokra Do you have sway to get webpack/webpack-sources#23 merged

**Comment by sompylasar** [comment]
According to https://github.com/webpack/webpack-sources/pull/23 , the fix for this issue has been merged 5 months ago (after over a year since submission) and released in `webpack-sources 1.2.0`.

`webpack` updates its dependency on `webpack-sources` from `1.0.1` to `1.2.0` in https://github.com/webpack/webpack/commit/cd9525ad4c6cea0a1e4f8011023098a591f9e0bc 

`webpack 4.17.2` released on Sep 3, 2018 contains the fix: https://github.com/webpack/webpack/releases/tag/v4.17.2
